File: ios/XiriGoEcommerce/Core/DesignSystem/Component/ShimmerModifier.swift
import SwiftUI

// MARK: - ShimmerModifier

/// Applies an animated shimmer gradient sweep to any view shape.
///
/// The linear gradient translates horizontally across the view on an infinite loop,
/// producing a loading-placeholder animation. All timing, angle, and color values
/// come from ``XGMotion/Shimmer`` design tokens -- no hardcoded values.
///
/// Usage:
/// ```swift
/// RoundedRectangle(cornerRadius: XGCornerRadius.medium)
///     .fill(XGMotion.Shimmer.gradientColors.first!)
///     .shimmerEffect()
/// ```
struct ShimmerModifier: ViewModifier {
    // MARK: - Internal

    /// When `false`, the modifier is a no-op and returns the content unchanged.
    let active: Bool

    func body(content: Content) -> some View {
        if active {
            content
                .overlay(shimmerOverlay)
                .mask(content)
        } else {
            content
        }
    }

    // MARK: - Private

    /// Multiplier for the total horizontal travel distance.
    private static let travelMultiplier: CGFloat = 2

    @State private var phase: CGFloat = 0

    private var shimmerOverlay: some View {
        GeometryReader { geometry in
            let gradientWidth = geometry.size.width
            LinearGradient(
                colors: XGMotion.Shimmer.gradientColors,
                startPoint: .leading,
                endPoint: .trailing,
            )
            .frame(width: gradientWidth)
            .rotationEffect(.degrees(XGMotion.Shimmer.angleDegrees))
            .offset(x: offsetForPhase(viewWidth: gradientWidth))
            .onAppear {
                withAnimation(
                    .linear(duration: XGMotion.Shimmer.duration)
                        .repeatForever(autoreverses: false),
                ) {
                    phase = 1
                }
            }
        }
        .clipped()
    }

    /// Translates the gradient from off-screen left to off-screen right.
    private func offsetForPhase(viewWidth: CGFloat) -> CGFloat {
        let totalTravel = viewWidth * Self.travelMultiplier
        return -viewWidth + (phase * totalTravel)
    }
}

// MARK: - View Extension

extension View {
    /// Applies an animated shimmer sweep effect for loading placeholders.
    ///
    /// - Parameter active: Pass `false` to disable the animation (renders content as-is).
    /// - Returns: The view with a shimmer gradient overlay when active.
    func shimmerEffect(active: Bool = true) -> some View {
        modifier(ShimmerModifier(active: active))
    }
}

// MARK: - PreviewConstants

private enum PreviewConstants {
    static let boxHeight: CGFloat = 120
    static let circleSize: CGFloat = 80
    static let textWidthLong: CGFloat = 220
    static let textWidthShort: CGFloat = 160
    static let textHeight: CGFloat = 16
}

// MARK: - ShimmerPreview

private struct ShimmerPreview: View {
    // MARK: - Internal

    var body: some View {
        VStack(spacing: XGSpacing.base) {
            rectanglePlaceholder
            circlePlaceholder
            textPlaceholders
        }
        .padding(XGSpacing.screenPaddingHorizontal)
    }

    // MARK: - Private

    private var rectanglePlaceholder: some View {
        RoundedRectangle(cornerRadius: XGCornerRadius.medium)
            .fill(XGMotion.Shimmer.gradientColors.first ?? XGColors.shimmer)
            .frame(height: PreviewConstants.boxHeight)
            .shimmerEffect()
    }

    private var circlePlaceholder: some View {
        Circle()
            .fill(XGMotion.Shimmer.gradientColors.first ?? XGColors.shimmer)
            .frame(
                width: PreviewConstants.circleSize,
                height: PreviewConstants.circleSize,
            )
            .shimmerEffect()
    }

    private var textPlaceholders: some View {
        VStack(alignment: .leading, spacing: XGSpacing.sm) {
            RoundedRectangle(cornerRadius: XGCornerRadius.small)
                .fill(XGMotion.Shimmer.gradientColors.first ?? XGColors.shimmer)
                .frame(
                    width: PreviewConstants.textWidthLong,
                    height: PreviewConstants.textHeight,
                )
                .shimmerEffect()

            RoundedRectangle(cornerRadius: XGCornerRadius.small)
                .fill(XGMotion.Shimmer.gradientColors.first ?? XGColors.shimmer)
                .frame(
                    width: PreviewConstants.textWidthShort,
                    height: PreviewConstants.textHeight,
                )
                .shimmerEffect()
        }
    }
}

// MARK: - Previews

#Preview("Shimmer on Shapes") {
    ShimmerPreview()
        .xgTheme()
}

#Preview("Shimmer Disabled") {
    RoundedRectangle(cornerRadius: XGCornerRadius.medium)
        .fill(XGMotion.Shimmer.gradientColors.first ?? XGColors.shimmer)
        .frame(height: PreviewConstants.boxHeight)
        .shimmerEffect(active: false)
        .padding(XGSpacing.screenPaddingHorizontal)
        .xgTheme()
}

File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/ShimmerModifier.kt
package com.xirigo.ecommerce.core.designsystem.component

import kotlin.math.tan
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.xirigo.ecommerce.core.designsystem.theme.XGCornerRadius
import com.xirigo.ecommerce.core.designsystem.theme.XGMotion
import com.xirigo.ecommerce.core.designsystem.theme.XGSpacing
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

/**
 * Applies an animated shimmer gradient sweep effect to any composable.
 *
 * The gradient translates horizontally across the component width on an infinite loop,
 * producing a loading-placeholder animation. Uses [XGMotion.Shimmer] design tokens for
 * duration, angle, and gradient colors — no hardcoded values.
 *
 * GPU-accelerated via [graphicsLayer] to ensure smooth 60 fps during scroll.
 *
 * @param enabled When `false` the modifier is a no-op (returns the chain unchanged).
 */
fun Modifier.shimmerEffect(enabled: Boolean = true): Modifier {
    if (!enabled) return this
    return composed {
        val transition = rememberInfiniteTransition(label = "shimmer")
        val translateAnimation by transition.animateFloat(
            initialValue = 0f,
            targetValue = 1f,
            animationSpec = infiniteRepeatable(
                animation = tween(
                    durationMillis = XGMotion.Shimmer.DURATION_MS,
                    easing = LinearEasing,
                ),
                repeatMode = RepeatMode.Restart,
            ),
            label = "shimmerTranslate",
        )

        val angleRadians = Math.toRadians(XGMotion.Shimmer.ANGLE_DEGREES.toDouble()).toFloat()

        graphicsLayer { }
            .drawWithContent {
                drawContent()

                val gradientWidth = size.width
                val tanAngle = tan(angleRadians)
                val yOffset = tanAngle * gradientWidth

                // Shift the gradient from off-screen left to off-screen right
                val translateX = translateAnimation * (size.width + gradientWidth) - gradientWidth

                val brush = Brush.linearGradient(
                    colors = XGMotion.Shimmer.GradientColors,
                    start = Offset(translateX, -yOffset),
                    end = Offset(translateX + gradientWidth, yOffset),
                )

                drawRect(brush = brush)
            }
    }
}

// region Preview

private val PreviewPlaceholderHeight = 16.dp
private val PreviewBoxHeight = 120.dp
private val PreviewCircleSize = 80.dp
private val PreviewTextWidthShort = 160.dp
private val PreviewTextWidthLong = 220.dp

@Preview(showBackground = true)
@Composable
private fun ShimmerEffectPreview() {
    XGTheme {
        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.Base),
        ) {
            // Shimmer on a rectangular box
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(PreviewBoxHeight)
                    .clip(RoundedCornerShape(XGCornerRadius.Medium))
                    .background(XGMotion.Shimmer.GradientColors.first())
                    .shimmerEffect(),
            )

            // Shimmer on a circle
            Box(
                modifier = Modifier
                    .size(PreviewCircleSize)
                    .clip(CircleShape)
                    .background(XGMotion.Shimmer.GradientColors.first())
                    .shimmerEffect(),
            )

            // Shimmer on text-width placeholders
            Column(verticalArrangement = Arrangement.spacedBy(XGSpacing.SM)) {
                Box(
                    modifier = Modifier
                        .width(PreviewTextWidthLong)
                        .height(PreviewPlaceholderHeight)
                        .clip(RoundedCornerShape(XGCornerRadius.Small))
                        .background(XGMotion.Shimmer.GradientColors.first())
                        .shimmerEffect(),
                )
                Box(
                    modifier = Modifier
                        .width(PreviewTextWidthShort)
                        .height(PreviewPlaceholderHeight)
                        .clip(RoundedCornerShape(XGCornerRadius.Small))
                        .background(XGMotion.Shimmer.GradientColors.first())
                        .shimmerEffect(),
                )
            }

            // Disabled shimmer (static)
            Spacer(modifier = Modifier.height(XGSpacing.SM))
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(PreviewPlaceholderHeight)
                    .clip(RoundedCornerShape(XGCornerRadius.Small))
                    .background(XGMotion.Shimmer.GradientColors.first())
                    .shimmerEffect(enabled = false),
            )
        }
    }
}

// endregion
